Feat/mac sandbox pre release pyinstaller

toofar opened the pull request:
<!-- Thanks for submitting a pull request! Please pick a descriptive title (not just "issue 12345"). If there is an open issue associated to your PR, please add a line like "Closes #12345" somewhere in the PR description (outside of this comment) -->

Switch to PyInstaller develop branch to get v6.0 changes which brings:

* WebEngine/Chromium sandboxing for macOS builds
* Removes the need for our downstream patching of the app on macOS
* Unifies resource lookups to be via importlib_resources (this is actually possible as of 2021 but we hadn't picked up on it)

This is a pretty standard adaptation to a new major version of a dependency except in this case the major version isn't released yet, so I'm pointing at a specific commit on their develop branch (the latest commit as of today).

In my opinion we probably want to do a little more testing of the resulting binaries before committing to this. Although I don't think the testing surface area is huge? Like it runs, resource files can be loaded, it can be signed? There are binaries [here](https://github.com/qutebrowser/qutebrowser/actions/runs/5678007969), based of [this](https://github.com/qutebrowser/qutebrowser/tree/feat/test_pre_release_pyinstaller) test branch with a custom CI job.

Potential testing checklist:

* [ ] installs over an old version with no warnings
* [ ] it launches
* [ ] can prompt to capture audio or location
    * test with jitsi meet
* [ ] :help is rendered like normal
* [ ] userscripts work (if packaged?)
* [ ] pdf-js works
    * https://www.ipcc.ch/site/assets/uploads/2021/06/Fact_sheet_AR6.pdf
    * or qute://pdfjs/build/pdf.js on builds with 5.15.2
* [ ] application icon on shortcut and taskbar is correct
* [ ] extensions are loaded
    * the reload command is registered


TODO:
* open issue to track reverting the temp stuff for the pre-release pyinstaller once a release comes out (requirements files and PYINSTALLER_COMPILE_BOOTLOADER) :heavy_check_mark:  #7806
* probably move the PYINSTALLER_COMPILE_BOOTLOADER=true out of tox into the build script so that it gets set for local builds too :heavy_check_mark: 
* do manual testing I guess. I went through a few earlier, but not all of them. :heavy_check_mark: 
* remove the warning page and the place that triggers it qutebrowser/html/warning-sandboxing.html :heavy_check_mark: 
* apologies to @rokm for pulling in the develop branch

Should I, GitHub, should I? :thinking: 
![image](https://github.com/qutebrowser/qutebrowser/assets/7419144/2cace295-151f-4c8d-8585-11a954f5dbfb)

Testing (it seems userscripts are not packaged):

Windows (10)

* [x] installs over an old version with no warnings
* [x] it launches
* [x] can prompt to capture audio or location
      * tested with jitsi meet
* [x] :help is rendered like normal
* [x] pdf-js works
      * https://www.ipcc.ch/site/assets/uploads/2021/06/Fact_sheet_AR6.pdf
* [x] application icon on shortcut and taskbar is correct
* [x] extensions are loaded
      * the reload command is registered

macOS (11.7.8)

* [x] installs over an old version with no warnings
* [x] it launches
* [x] can prompt to capture audio or location
      * tested with jitsi meet
* [x] :help is rendered like normal
* [x] pdf-js works
      * https://www.ipcc.ch/site/assets/uploads/2021/06/Fact_sheet_AR6.pdf
* [x] application icon on shortcut and taskbar is correct
* [x] extensions are loaded
      * the reload command is registered

Closes: #7278

toofar commented on misc/requirements/requirements-pyinstaller.txt-raw:
```diff
@@ -1 +1 @@
-PyInstaller
+git+https://github.com/pyinstaller/pyinstaller.git@develop
```
Not sure this is the right thing to be doing here. I want the dependency update robot to notify us when the new release is out but not to upgrade to a new arbitrary commit on the branch. I don't think this will do that. Probably I should just put the commit sha here and subscribe to releases on the repo.

The-Compiler replied:
Given that the selection of what commit to use is somewhat arbitrary, personally I'd be fine with updating this in a controlled fashion once a week (and reviewing what changed during the dependency upgrade) until PyInstaller 6 is out.

toofar replied:
I'm not sure I agree with that. Because it's an arbitrary commit I expect us to have to do extra verification of it as it hasn't been through upstream's normal release process. Are the smoke tests good enough to cover all the functionality the installer could break? I don't know but I would rather only do the extra verification I think is required once.
They seem to release every month so I doubt we would be waiting long.

toofar replied:
I've pinned this to a specific commit in both requirements files for now. Let me know if you really want me to put it back (recompile_requirements.py was updating to the latest develop commit just fine with the version commented on here).

The-Compiler commented on scripts/dev/build_release.py:
```diff
@@ -342,7 +293,6 @@ def build_mac(
     utils.print_title("Building .app via pyinstaller")
     call_tox(f'pyinstaller-64bit{"-qt5" if qt5 else ""}', '-r', debug=debug)
     utils.print_title("Patching .app")
```
nitpick: Should probably be removed too?

The-Compiler commented on scripts/dev/build_release.py:
```diff
@@ -342,7 +293,6 @@ def build_mac(
     call_tox(f'pyinstaller-64bit{"-qt5" if qt5 else ""}', '-r', debug=debug)
     utils.print_title("Patching .app")
-    patch_mac_app(qt5=qt5)
     utils.print_title("Re-signing .app")
     sign_mac_app()
```
I think we can get rid of the re-signing now that we don't patch anymore? Could you try to remove the re-signing part from `sign_mac_app` (so that it only verifies the signature), perhaps rename it to something more appropriate, and see what happens?

toofar replied:
![image](https://github.com/qutebrowser/qutebrowser/assets/7419144/f988f765-f867-4a51-9bcc-bd1f2d7c6b08)

Seems to work fine without, based on my extremely limited knowledge. It installed (drag over Applications shortcut after opening .dmg) over top of a previously nightly fine, and ran. Using codesign (-dv and -vv) on the application in a terminal yields the same result before and after. (The .dmg isn't signed, which seems to be a thing you can do these days.)

toofar commented:
I've pushed again to:
1. remove the re-signing
2. move the PYINSTALLER_COMPILE_BOOTLOADER env var setting into build_release instead of tox.ini so that if you run build_release not via tox you don't have to remember to set it. It's a bit ugly, and while we are going to remove that stuff after the new pyinstaller release comes out, let me know if you do manual builds via tox anyway and I might move it back to how it was before

While looking through the spec file I saw [this note](https://github.com/qutebrowser/qutebrowser/blob/ffc06e5/misc/qutebrowser.spec#L140) and that the comment it refers to was removed a couple of years ago in https://github.com/pyinstaller/pyinstaller/commit/af23dcceb3f997bda40. So it might not be needed anymore.

I'll update the PR description with potential testing notes based off of looking through the spec file.

rokm commented:
> While looking through the spec file I saw [this note](https://github.com/qutebrowser/qutebrowser/blob/ffc06e5/misc/qutebrowser.spec#L140) and that the comment it refers to was removed a couple of years ago in [pyinstaller/pyinstaller@af23dcceb3f997bda40](https://github.com/pyinstaller/pyinstaller/commit/af23dcceb3f997bda40). So it might not be needed anymore.

This looks like a work-around for PyInstaller butchering the `QtWebEngineCore.framework` bundle.

It should be safe to replace the bundle ID with your actual bundle ID now that the .framework bundles are preserved.

toofar commented:
I updated the bundle ID and went through my manual testing stuff and everything seemed to work :tada:. Although I'm not sure how I'm supposed to reproduce the original failure, did it just stop the app from starting?

Also realized there is the macOS sandbox warning page which is probably unneeded now? Or at least should have references to the PyInstaller issue removed.

More interestingly the extensions are no longer being loaded on macOS. Resulting in messages like: "scroll-to-perc: no such command" and "reload: no such command".

```
>>> from qutebrowser.extensions import loader
>>> list(loader.walk_components())
[]

>> components.__path__
['/Applications/qutebrowser.app/Contents/Frameworks/qutebrowser/components']
>> components.__name__
'qutebrowser.components'

>> pkgutil.iter_importers('qutebrowser.components')
[FileFinder(...)]
```

However `/Applications/qutebrowser.app/Contents/Frameworks/qutebrowser/components` is not a real path (the parent folder does exist, it's just got data files in it though).

The workaround of using the PyInstaller importer toc manually (removed in https://github.com/qutebrowser/qutebrowser/commit/dd2fc8e10bb9d4a1bd0158110173793a18736d6b) seems like it'll still work though:

```
>>> {p for p in list(pkgutil.iter_importers('qutebrowser'))[2].toc if p.startswith('qutebrowser.components')}
{'qutebrowser.components',
'qutebrowser.components.adblockcommands',
'qutebrowser.components.braveadblock',
'qutebrowser.components.caretcommands',
'qutebrowser.components.hostblock',
'qutebrowser.components.misccommands',
'qutebrowser.components.readlinecommands',
'qutebrowser.components.scrollcommands',
'qutebrowser.components.utils',
'qutebrowser.components.utils.blockutils',
'qutebrowser.components.zoomcommands'}
```

The-Compiler commented:
Thanks for all the great work, this definitely gets us a big step closer! <3

> I've pinned this to a specific commit in both requirements files for now.

Works for me.

> While looking through the spec file I saw [this note](https://github.com/qutebrowser/qutebrowser/blob/ffc06e5/misc/qutebrowser.spec#L140) and that the comment it refers to was removed a couple of years ago in https://github.com/pyinstaller/pyinstaller/commit/af23dcceb3f997bda40. So it might not be needed anymore.

Nice catch!

> let me know if you do manual builds via tox anyway and I might move it back to how it was before

I do indeed to ensure I have a consistent environment with everything it needs. So I suppose dd86483065a81726e3ce70ec9fab7389c12efeb4 and 8a327beb17dbe97cc8022549d652f72eb2996c97 could be reverted?

> Although I'm not sure how I'm supposed to reproduce the original failure, did it just stop the app from starting?

I believe renderer processes just didn't start, yeah.

> Also realized there is the macOS sandbox warning page which is probably unneeded now? Or at least should have references to the PyInstaller issue removed.

Yep, reverting that was the right call IMHO!

> More interestingly the extensions are no longer being loaded on macOS. Resulting in messages like: "scroll-to-perc: no such command" and "reload: no such command".

That's weird! Might be good to try and see if we can get a more minimal reproducer to report this as a PyInstaller regression? https://github.com/pyinstaller/pyinstaller/pull/5959 comes with a testcase which [hasn't had any changes](https://github.com/pyinstaller/pyinstaller/blob/develop/tests/functional/test_pkgutil.py) and still seems to pass, so I'm not sure what it is that we do differently?

toofar commented:
Alright, last call for comments on this one! Latest changes:

* I added a Windows 10 minimum version check to the installer
    * based on @bitraid's comment here https://github.com/qutebrowser/qutebrowser/pull/7315#issuecomment-1657094825
    * but with an added conditional to do the old check for the Qt5 installer
    * and I naughtily bumped the Qt6 OS requirements down a bit from Win 10 build 1809 -> build 1607 based on our testing. This is lower than the Qt advice but seems to work. Easier enough to increase it again if someone finds an issue?
* dropped the two ugly commits around moving the pyinstaller bootload compiling stuff out of tox into the build_release script
* adaptions from #7804 
* I added the Qt5 builds back to my test CI job and actually tested them too
    * latest binaries are here: https://github.com/qutebrowser/qutebrowser/actions/runs/5839589092
    * I had one fright where installing the windows Qt5 build over top of the Qt6 build causes the application to fail to launch with a stack trace about failing to import PyQt6.QtCore. It seems there was a leftover, empty, `_internal\PyQt6\Qt6\bin\locales` folder left in the install dir that was tricking the autoselect logic. I expect doing an uninstall first would resolve this, as that should remove the whole install dir
    * pdfjs doesn't work on the binary Qt5 version we are packaging, as probably noted elsewhere. Changed the testing instructions to just check that `qute://pdfjs/build/pdf.js` returns JS
* I didn't update `scripts/dev/download_release.sh`, easy enough to do that during the release
    
Too old win10 warning (before I dropped the requirement lower):

![image](https://github.com/qutebrowser/qutebrowser/assets/7419144/72be9518-9fbe-42ed-aabf-9ade57d4f7a0)

Error dialog on launch after installing the Qt5 version over the Qt6 version:

![image](https://github.com/qutebrowser/qutebrowser/assets/7419144/e22160e0-9c3a-42be-aaf6-44216765538f)

Testing the autoselect logic with (after starting with `--qt-wrapper PyQt5`):

![image](https://github.com/qutebrowser/qutebrowser/assets/7419144/d06a9e17-ff71-4961-9d27-b341aa7f5312)

Tomorrow I'll look at doing a new test case for pyinstaller and move to main my changes around getting better smoke test erorr logs, but that's all the stuff I have planned for this PR.

The-Compiler commented on misc/nsis/qutebrowser.nsi:
@bitraid could you give this a look?

The-Compiler commented on misc/nsis/install.nsh:
@bitraid could you give this a look?

bitraid replied:
I think it would be better to have the QT5 be a compile time check:

```nsis
; OS version check
  !if "${QT5}" == "True"
    !insertmacro CheckPlatform ${PLATFORM}
    !insertmacro CheckMinWinVer ${MIN_WIN_VER}
    Goto _os_check_pass
  !else
    ${If} ${RunningX64}
      GetWinVer $R0 Major
      IntCmpU $R0 10 0 _os_check_fail _os_check_pass
      GetWinVer $R1 Build
      ${If} $R1 >= 22000 ; Windows 11 21H2
        Goto _os_check_pass
      ${ElseIf} $R1 >= 14393 ; Windows 10 1607
      ${AndIf} ${IsNativeAMD64} ; Windows 10 has no x86_64 emulation on arm64
        Goto _os_check_pass
      ${EndIf}
    ${EndIf}
    _os_check_fail:
    MessageBox MB_OK|MB_ICONSTOP "This version of ${PRODUCT_NAME} requires a 64-bit$\r$\n\
      version of Windows 10 1607 or later."
    Abort
  !endif
  _os_check_pass:
```

Although, on 32-bit systems, the `CheckPlatform` macro will give the user the somewhat misleading message of `Please, run the 32-bit installer of ${PRODUCT_NAME} on this version of Windows.`. Maybe it's better to no longer use the `CheckPlatform`/`CheckMinWinVer` macros of NsisMultiUser plugin:

```nsis
; OS version check
  ${If} ${RunningX64}
    GetWinVer $R0 Major
    !if "${QT5}" == "True"
      IntCmpU $R0 6 0 _os_check_fail _os_check_pass
      GetWinVer $R1 Minor
      IntCmpU $R1 2 _os_check_pass _os_check_fail _os_check_pass
    !else
      IntCmpU $R0 10 0 _os_check_fail _os_check_pass
      GetWinVer $R1 Build
      ${If} $R1 >= 22000 ; Windows 11 21H2
        Goto _os_check_pass
      ${ElseIf} $R1 >= 14393 ; Windows 10 1607
      ${AndIf} ${IsNativeAMD64} ; Windows 10 has no x86_64 emulation on arm64
        Goto _os_check_pass
      ${EndIf}
    !endif
  ${EndIf}
  _os_check_fail:
  !if "${QT5}" == "True"
    MessageBox MB_OK|MB_ICONSTOP "This version of ${PRODUCT_NAME} requires a 64-bit$\r$\n\
      version of Windows 8 or later."
  !else
    MessageBox MB_OK|MB_ICONSTOP "This version of ${PRODUCT_NAME} requires a 64-bit$\r$\n\
      version of Windows 10 1607 or later."
  !endif
  Abort
  _os_check_pass:
```

The-Compiler replied:
Personally I don't think we'll need to perfect this - it's only to be quite temporary until your new installer is merged :slightly_smiling_face:  I'll let @toofar make the call on this, happy to keep it as-is as well.

toofar replied:
Having a more accurate error message sounds worthwhile to me, applied.

The-Compiler commented on misc/requirements/requirements-pyinstaller.txt-raw:
```diff
@@ -1 +1 @@
-PyInstaller
+pyinstaller @ git+https://github.com/pyinstaller/pyinstaller.git@79f62ef29822169ae00cd4271390d0e3175476ad
```
If you plan on doing another round of testing with your latest changes, maybe worth bumping up this again? If you'd rather not, I'm fine with that too.

toofar replied:
I've done more testing but I don't plan on updating the commit. I've been checking in on subsequent commits and I don't see anything relevant to us.

The-Compiler commented on qutebrowser/extensions/loader.py:
```diff
@@ -93,6 +102,23 @@ def walk_components() -> Iterator[ExtensionInfo]:
+def _walk_pyinstaller() -> Iterator[ExtensionInfo]:
+    """Walk extensions when using PyInstaller.
+
+    See https://github.com/pyinstaller/pyinstaller/issues/1905
```
Note to self, we still should try to figure out whether this has regressed.

toofar replied:
I don't think it has, but I'm not sure if we can test it on an old version because of the workarounds we had to do symlinking stuff around. I suppose this might be an issue with the upstream fix for that stuff breaking this use case.

This only started happening with https://github.com/qutebrowser/qutebrowser/pull/7803/commits/c0d31d399d31d3c45f47e375826ed87c16807edf where the data files changed to be under `/Frameworks/qutebrowser/`. That's the same path that the python files are reported to be under so the issue is that somehow the `/Frameworks/qutebrowser/` folder existing is causing importlib to use the file loader instead of the frozen ones. But it seems to work fine on Windows (TODO: triple check with https://github.com/qutebrowser/qutebrowser/actions/runs/5846907490), so it might be some mac specific workaround going wonky.

I tried to make an upstream test fail [here](https://github.com/toofar/pyinstaller/pull/1/files), but it didn't. I think the script in that test as the right structure (a folder with data files at the same level as a folder with python files). I might switch tack and try to do a minimal reproducer based on qutebrowser (including spec file).

Anyway, I don't intend to treat this oddity as blocking (I really want to see all the nightly tests working again!). So I'll probably merge this PR tomorrow.

rokm replied:
This might be fixed by https://github.com/pyinstaller/pyinstaller/pull/7885.

toofar replied:
Looks like that fixes it indeed. Data files and dynamic modules are both being loaded fine with no special casing for frozen builds from us! Yet another issue that gets solved with me doing nothing but procrastinating :joy: 

I just did a quick build to check. Will double check in more detail tomorrow.

The-Compiler commented:
As for the "empty PyQt6 dir" thing: I've seen that too, and considered changing the logic to account for that - but then decided this is one of the "user's environment is in a weird inconsistent thing" situations I'd much rather report to the user rather than just painting over it and assuming something.

bitraid commented on misc/nsis/qutebrowser.nsi:
```diff
@@ -124,11 +124,16 @@ ShowUninstDetails hide
+!endif
+
+; If not defined, assume Qt6 (more requires a more recent windows version)
+!ifndef QT5
+  !define QT5 "False"
 !endif
```
Could be one line:
```nsis
!define /ifndef QT5 "False"
```
